app/lib/account/backend.dart

```diff
@@ -180,6 +180,47 @@ class AccountBackend {
     return user;
   }
 
+  /// Returns `true` if the given [user] likes the package named [package].
+  /// Otherwise returns `false`.
+  Future<bool> getPackageLikeStatus(User user, String package) async {
+    final key =
+        _db.emptyKey.append(User, id: user.id).append(Like, id: package);
+    try {
+      await _db.lookup([key]);
+    } on KeyNotFoundException {
+      return false;
+    }
+    return true;
+  }
+
+  /// Return an iterable with the names of all the packages that the given [user] likes.
+  Future<Iterable<String>> listPackagesLikes(User user) async {
+    final likes = await _db.query<Like>(ancestorKey: user.key).run().toList();
+    return likes.map((Like l) => l.package);
+  }
+
+  /// Creates an a package like entry for the given [user] and [package].
+  Future<void> putPackageLike(User user, String package) async {
+    final newLike = Like()
+      ..parentKey = user.key
+      ..id = package;
+    newLike.created = DateTime.now().toUtc();
+    await _db.commit(inserts: [newLike]);
```

Comment:
Okay, next big step is creating a `Package.likes` property in `package/models.dart`, and then create a transaction here.

In this transaction we'll have to do the following:
 * Look to see if the `Like` entity already exists (if it does, abort transaction and return),
 * Fetch the package entity, and increment the `likes` count,
 * Create a new `Like` entity
 * Commit both the package entity and `Like` entity in the same transaction commit call.

Similarly, we'll have to do that kind of lookup when we do `unlike` operation, to ensure that we decrement and delete in the same transaction, thus, we don't decrement unless we also removed the [Like] entity.

Note. that for this it might be simpler to move the `lookupPackage` call in here:
```
  final p = await packageBackend.lookupPackage(package);
  if (p == null) {
    throw NotFoundException.resource(package);
  }
```
If this method throws `NotFoundException` that's fine as it'll just get caught higher up and turned into a nice error message :)

-----
If we do the `Package.likes` update on-the-fly we limit ourselves to one transaction per second per package. Which is probably fine. We can retry it, and like-congestion for a package is so high that we can't successfully like the package we just fail the operation. This is not idea, but it'll be very very rare, and it's a sign that the feature is very very popular -- so crazy idea: If with 8 retries with exponential-backoff and randomized intervals (see `package:retry`) we can't successfully like the package because of like-congestion -- we print to the logs a warning: `"[alert:like-congestion] $package is too likable, liking-transaction failed 8 consecutive attempts -- please rework this code-path."`.

Then in GCP logging console we can create a custom metrics that measures number of times we see a log message containing `[alert:like-congestion]`. And in stackdriver monitoring we can create an alert that sends us an email if this ever happens. If it just happens once every 6 months, I don't care; if it happens frequently we can prioritize it.

We can also solve it with a background job, but I'm not sure we need this upfront. It might be better to do that later, if the this problem ever arises.